internal/cli/server_upgrade.go

```diff
@@ -0,0 +1,363 @@
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/posener/complete"
+
+	"github.com/hashicorp/waypoint-plugin-sdk/terminal"
+	clientpkg "github.com/hashicorp/waypoint/internal/client"
+	"github.com/hashicorp/waypoint/internal/clierrors"
+	"github.com/hashicorp/waypoint/internal/clisnapshot"
+	"github.com/hashicorp/waypoint/internal/pkg/flag"
+	pb "github.com/hashicorp/waypoint/internal/server/gen"
+	"github.com/hashicorp/waypoint/internal/serverclient"
+	"github.com/hashicorp/waypoint/internal/serverinstall"
+)
+
+type ServerUpgradeCommand struct {
+	*baseCommand
+
+	platform     string
+	contextName  string
+	snapshotName string
+	skipSnapshot bool
+	confirm      bool
+}
+
+func (c *ServerUpgradeCommand) initWriter(fileName string) (io.Writer, io.Closer, error) {
+	f, err := os.Create(fileName)
+
+	if err != nil {
+		return nil, nil, err
+	} else {
+		return f, f, nil
+	}
+}
+
+func (c *ServerUpgradeCommand) Run(args []string) int {
+	ctx := c.Ctx
+	log := c.Log.Named("upgrade")
+	defer c.Close()
+
+	// Initialize. If we fail, we just exit since Init handles the UI.
+	if err := c.Init(
+		WithArgs(args),
+		WithFlags(c.Flags()),
+		WithNoConfig(),
+	); err != nil {
+		return 1
+	}
+
+	// Error handling from input
+
+	if !c.confirm {
+		c.ui.Output(confirmReqMsg, terminal.WithErrorStyle())
+		return 1
+	}
+
+	if c.platform == "" {
+		c.ui.Output(
+			"A platform is required and must match the server context",
+			terminal.WithErrorStyle(),
+		)
+		return 1
+	}
+
+	p, ok := serverinstall.Platforms[strings.ToLower(c.platform)]
+	if !ok {
+		c.ui.Output(
+			"Error upgrading server on %s: invalid platform",
+			c.platform,
+			terminal.WithErrorStyle(),
+		)
+
+		return 1
+	}
+
+	// Finish error handling
+
+	// Get Server config to preserve existing configurations from context
+	var ctxName string
+	if c.contextName != "" {
+		ctxName = c.contextName
+	} else {
+		defaultName, err := c.contextStorage.Default()
+		if err != nil {
+			c.ui.Output(
+				"Error getting default context: %s",
+				clierrors.Humanize(err),
+				terminal.WithErrorStyle(),
+			)
+			return 1
+		}
+		ctxName = defaultName
+	}
+
+	originalCfg, err := c.contextStorage.Load(ctxName)
+	if err != nil {
+		c.ui.Output(
+			"Error loading the context %q: %s",
+			ctxName,
+			clierrors.Humanize(err),
+			terminal.WithErrorStyle(),
+		)
+		return 1
+	}
+
+	// Upgrade waypoint server
+	sg := c.ui.StepGroup()
+	defer sg.Wait()
+
+	s := sg.Add("Validating server context: %q", ctxName)
+	defer func() { s.Abort() }()
+
+	conn, err := serverclient.Connect(ctx, serverclient.FromContextConfig(originalCfg))
+	if err != nil {
+		c.ui.Output(
+			"Error connecting with context %q: %s",
+			ctxName,
+			clierrors.Humanize(err),
+			terminal.WithErrorStyle(),
+		)
+		return 1
+	}
+
+	s.Update("Verifying connection is valid for context %q...", ctxName)
+
+	client := pb.NewWaypointClient(conn)
+	if _, err := clientpkg.New(ctx,
```

Comment:
I'd like to see a comment explaining why we initialize `clientpkg` if we don't use it. 

(It is important, we have to verify API compat, but I'd put a comment at least saying that)

Reply:
Ah yeah I'll leave a note here explaining this. Although perhaps the client getting version info on 147 is enough to verify API compatibility?

Reply:
It isn't! You have to do the whole "negotiation" dance which clientpkg does. :).

Reply:
Ok perfect! I'll write up a note then so it's more clear :) thanks!